### PR TITLE
kiwitalk: init at 0.5.1

### PR DESCRIPTION
**Pull request by honnip**
## Description of changes
Init https://github.com/KiwiTalk/KiwiTalk
KiwiTalk is a cross-platform client of KakaoTalk.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    install -DT backend/bin/app/icons/32x32.png $out/share/icons/hicolor/32x32/apps/kiwi-talk.png
+  '';
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "KiwiTalk";
+      exec = "kiwi-talk";
+      icon = "kiwi-talk";
+      desktopName = "KiwiTalk";
+      comment = meta.description;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ki/kiwitalk/package.nix:116]:
```suggestion
      comment = "An UNOFFICIAL cross-platform KakaoTalk client";
```
a change in meta should not trigger a rebuild

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  version = "0.5.1";
+
+  src = fetchFromGitHub {
+    owner = "KiwiTalk";
+    repo = "KiwiTalk";
+    rev = "v${version}";
+    hash = "sha256-Th8q+Zbc102fIk2v7O3OOeSriUV/ydz60QwxzmS7AY8=";
+  };
+
+  sourceRoot = src.name;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ki/kiwitalk/package.nix:31]: Isn't this the default?

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+      rm -rf $out/v3/tmp
+      for f in $(find $out -name "*.json"); do
+        sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+        jq --sort-keys . $f | sponge $f
+      done
+    '';
+
+    dontBuild = true;
+    dontFixup = true;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ki/kiwitalk/package.nix:62]: I think we could replace the mkDerivation with runCommand, can't we?

**Comment by honnip** [comment]
Cannot build (`pnpm install --frozen-lockfile --ignore-script` part) with runCommand due to permissions issues.  

log:
```log
 EACCES  EACCES: permission denied, mkdir '/nix/store/dfbrs9s5786xzlfiwy2nwclg5sf299xy-source/node_modules'
```

**Comment by SuperSandro2000** [comment]
Then we just leave it as is :)

**Review by SuperSandro2000 — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by pbsds** [comment]
Both i and ofborg seem to struggle fetching the pnpm deps:

```
┃        > FetchError: request to https://registry.npmjs.org/@vanilla-extract/private/-/private-1.0.3.tgz failed, reason: unable to get local issuer certificate
┃        >     at ClientRequest.<anonymous> (/nix/store/y3mxy318cr8cl26saysfwqbxbcwb31lb-pnpm-8.10.5/lib/node_modules/pnpm/dist/pnpm.cjs:73734:18)
┃        >     at ClientRequest.emit (node:events:514:28)
┃        >     at TLSSocket.socketErrorListener (node:_http_client:495:9)
┃        >     at TLSSocket.emit (node:events:526:35)
┃        >     at emitErrorNT (node:internal/streams/destroy:151:8)
┃        >     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
┃        >     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
┃        > Progress: resolved 925, reused 0, downloaded 0, added 0
```

**Comment by honnip** [comment]
Fixed. Thank you

**Comment by pbsds** [comment]
There is a hash mismatch on aarch64-linux

```
error: hash mismatch in fixed-output derivation '/nix/store/ffjdgad8fznc5wvagxpxda48ma5wc7gs-kiwitalk-pnpm-deps-0.5.1.drv':
         specified: sha256-LJPjWNpVfdUu8F5BMhAzpTo/h6ax7lxY2EESHj5P390=
            got:    sha256-N1K4pV5rbWmO/KonvYegzBoWa6TYQIqhQyxH/sWjOJQ=
```

I tried building locally on x86_64-linux which succeeded, so the mismatch is not due to timing.

Does pnpm fetch platform-specific binaries or artifacts?

**Comment by honnip** [comment]
Yes, there are multiple versions of esbuild as dependencies. I don't know if there is a better way other than hashing each time.

**Comment by pbsds** [comment]
Then consider something lie https://github.com/aacebedo/nixpkgs/blob/c747c62364360316242dc87309e04f7cacbd1ba6/pkgs/applications/audio/youtube-music/default.nix#L55-L60

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      categories = [ "Network" "InstantMessaging" ];
+    })
+  ];
+
+  meta = with lib; {
+    description = "An UNOFFICIAL cross-platform KakaoTalk client written in TypeScript & Rust (SolidJS, tauri)";
+    homepage = "https://github.com/KiwiTalk/KiwiTalk";
+    maintainers = with maintainers; [ honnip ];
+    license = licenses.asl20;
+    platforms = platforms.linux;
```
Comment (pbsds) [inline pkgs/by-name/ki/kiwitalk/package.nix:131]: Considering you provide darwin hashes, should this be relaxed to
```suggestion
    platforms = platforms.linux ++ platforms.darwin;
```

?

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      jq
+      moreutils
+      nodePackages.pnpm
+      cacert
+    ];
+
+    installPhase = ''
+      export HOME=$(mktemp -d)
+      pnpm config set store-dir $out
+      # cat package.json | jq '.pnpm.supportedArchitectures += { "os": ["linux"], "cpu": ["arm64"] }' | sponge package.json
```
Comment (pbsds) [inline pkgs/by-name/ki/kiwitalk/package.nix:51]: Is this how you derived the hashes? If so please add a comment indicating how and why to use it.

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      categories = [ "Network" "InstantMessaging" ];
+    })
+  ];
+
+  meta = with lib; {
+    description = "An UNOFFICIAL cross-platform KakaoTalk client written in TypeScript & Rust (SolidJS, tauri)";
+    homepage = "https://github.com/KiwiTalk/KiwiTalk";
+    maintainers = with maintainers; [ honnip ];
+    license = licenses.asl20;
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (pbsds) [inline pkgs/by-name/ki/kiwitalk/package.nix:135]:
```suggestion
    platforms = platforms.linux ++ platforms.darwin;
    mainProgram = "kiwi-talk";
```

File: pkgs/by-name/ki/kiwitalk/package.nix [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+  preInstall = ''
+    mv target/release/bundle/deb/*/data/usr/ $out
+  '';
+
+  postInstall = ''
+    install -DT backend/bin/app/icons/512x512.png $out/share/icons/hicolor/512x512/apps/kiwi-talk.png
+    install -DT backend/bin/app/icons/256x256.png $out/share/icons/hicolor/256x256/apps/kiwi-talk.png
+    install -DT backend/bin/app/icons/128x128.png $out/share/icons/hicolor/128x128/apps/kiwi-talk.png
+    install -DT backend/bin/app/icons/32x32.png $out/share/icons/hicolor/32x32/apps/kiwi-talk.png
```
Comment (pbsds) [inline pkgs/by-name/ki/kiwitalk/package.nix:116]:
```suggestion
    install -Dm644 -T backend/bin/app/icons/512x512.png $out/share/icons/hicolor/512x512/apps/kiwi-talk.png
    install -Dm644 -T backend/bin/app/icons/256x256.png $out/share/icons/hicolor/256x256/apps/kiwi-talk.png
    install -Dm644 -T backend/bin/app/icons/128x128.png $out/share/icons/hicolor/128x128/apps/kiwi-talk.png
    install -Dm644 -T backend/bin/app/icons/32x32.png $out/share/icons/hicolor/32x32/apps/kiwi-talk.png
```

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 269182` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kiwitalk</li>
  </ul>
</details>

Looks mostly good to me, just 

1. the icons seem to have the executable bit set, see the code suggestion 
2. there is a duplicate desktop entry:

```
[nix-shell:/dev/shm/nixpkgs-review/pr-269182-1]$ cat results/kiwitalk/share/applications/kiwi-talk.desktop
[Desktop Entry]
Categories=Network;
Exec=kiwi-talk
Icon=kiwi-talk
Name=KiwiTalk
Terminal=false
Type=Application

[nix-shell:/dev/shm/nixpkgs-review/pr-269182-1]$ cat results/kiwitalk/share/applications/KiwiTalk.desktop
[Desktop Entry]
Categories=Network;InstantMessaging
Comment=An UNOFFICIAL cross-platform KakaoTalk client
Exec=kiwi-talk
Icon=kiwi-talk
Name=KiwiTalk
Type=Application
Version=1.4
```

can the one you don't make in the derivation be removed?

**Comment by honnip** [comment]
I was missing lots of things 🥲  


`cargo-tauri` also exported a desktop entry and icons, so I have removed the generated entry and don't need to `install` icons.

**Comment by pbsds** [comment]
The darwin versions doesn't currently build since the build input `libayatana-appindicator` does not support darwin. Likely it is a platform-specific input and is not at all needed on darwin, but this I think is for a darwinista figure out and tests.

LGTM, thanks you!

**Pull request merged**